python/ray/serve/controller.py

```diff
@@ -249,6 +249,26 @@ async def deploy(self, name: str, backend_config: BackendConfig,
                      replica_config: ReplicaConfig,
                      version: Optional[str]) -> Optional[GoalId]:
         """TODO."""
+
+        # Infer HTTP route and methods based on replica_config.
+        if replica_config.path_prefix is None:
+            replica_config.path_prefix = f"/{name}"
+            # Updated here because it is used by backend worker
+            backend_config.internal_metadata.path_prefix = f"/{name}"
+        route = replica_config.path_prefix
+        methods = ["GET", "POST"]
+        if replica_config.is_asgi_app:
+            # When the backend is asgi application, we want to proxy it
+            # with a prefixed path as well as proxy all HTTP methods.
+            # {wildcard:path} is used so HTTPProxy's Starlette router can match
+            # arbitrary path.
+            route = "/{route}/{wildcard:path}"
+            # https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
+            methods = [
+                "GET", "HEAD", "POST", "PUT", "DELETE", "CONNECT", "OPTIONS",
+                "TRACE", "PATCH"
+            ]
```

Comment:
I'm a bit confused at this logic overall. Shouldn't there only be a single branch here?
```python
if replica_config.is_ingress:
    route = replica_config.path_prefix or f"{name}"
    methods = [...] # all
```

Reply:
we still need to set the default path prefix using the backend name, updated the code to make it cleaner.